app/controllers/api/groups_controller.rb

```diff
@@ -401,6 +401,62 @@ def collect_submission
         HttpStatusHelper::ERROR_CODE['message']['201'] }, status: :created
     end
 
+    def add_test_results
+      begin
+        raw_body = request.body.read
+
+        # Enforce size limit
+        if raw_body.bytesize > Settings.upload.max_body_size.to_i
+          return render json: { errors: 'Request body too large' }, status: :content_too_large
+        end
+
+        payload = JSON.parse(raw_body)
+      rescue JSON::ParserError => e
+        return render json: { errors: "Invalid JSON: #{e.message}" }, status: :bad_request
+      end
+
+      unless payload.is_a?(Hash) && payload.key?('test_results')
+        return render json: { errors: 'Missing required key: test_results' }, status: :bad_request
+      end
+
+      # Validate test results against contract schema
+      validation = TestResultsContract.new.call(payload['test_results'])
+
+      if validation.failure?
+        return render json: { errors: validation.errors.to_hash }, status: :unprocessable_content
+      end
+
+      # Verify grouping exists and is authorized (grouping helper method handles this)
+      if grouping.nil?
+        return render json: { errors: 'Group not found for this assignment' }, status: :not_found
+      end
+
+      # Verify submission exists before attempting to create test run
+      submission = grouping.current_submission_used
+      if submission.nil?
+        return render json: { errors: 'No submission exists for this grouping' }, status: :unprocessable_content
+      end
+
+      begin
+        ActiveRecord::Base.transaction do
+          test_run = TestRun.create!(
+            status: :in_progress,
+            role: current_role,
+            grouping: grouping,
+            submission: submission
+          )
+
+          test_run.update_results!(payload['test_results'])
+          render json: { status: 'success', test_run_id: test_run.id }, status: :created
+        end
+      rescue ActiveRecord::RecordInvalid => e
+        render json: { errors: e.record.errors.full_messages }, status: :unprocessable_content
+      rescue StandardError => e
+        Rails.logger.error("Test results processing failed: #{e.message}\n#{e.backtrace.join("\n")}")
```

Comment:
The logging is good, but we're using a custom logger class (see examples using `m_logger`/`MarkusLogger.instance`)